pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/utils/clock"
+)
+
+// PodStartupLatencyTracker records key moments for startup latency calculation,
+// e.g. image pulling or pod observed running on watch.
+type PodStartupLatencyTracker struct {
+	// protect against concurrent read and write on pods map
+	lock sync.Mutex
+	pods map[types.UID]*perPodState
+	// For testability
+	clock clock.Clock
+}
+
+type perPodState struct {
+	firstStartedPulling                         time.Time
+	lastFinishedPulling                         time.Time
+	firstResourceVersionWhenPodWasSeenAsRunning uint64
+	// log, if pod latency was already Observed
+	recorded bool
+}
+
+func NewPodStartupLatencyTracker() *PodStartupLatencyTracker {
+	return &PodStartupLatencyTracker{
+		pods:  map[types.UID]*perPodState{},
+		clock: clock.RealClock{},
+	}
+}
+
+// ObservedPodOnWatch to be called from somewhere where we look for pods.
+func (p *PodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when time.Time) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		p.pods[pod.UID] = &perPodState{}
+		return
+	}
+
+	if state.firstResourceVersionWhenPodWasSeenAsRunning == 0 {
+		// skip, pod didn't start yet
+		return
+	}
+	if state.recorded {
+		// skip, pod's latency already recorded
+		return
+	}
+
+	podResourceVersion, err := strconv.ParseUint(pod.ResourceVersion, 10, 64)
+	if err != nil {
+		klog.ErrorS(err, "Unable to parse", "pod.ResourceVersion", pod.ResourceVersion)
+		return
+	}
+
+	if podResourceVersion >= state.firstResourceVersionWhenPodWasSeenAsRunning {
+		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
+		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
+		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
+
+		klog.V(2).InfoS("Observe PodStartSLIDuration metric", "pod", klog.KObj(pod), "podStartSLOduration", podStartSLOduration, "watchObservedTime", when, "pod.CreationTimestamp", pod.CreationTimestamp.Time, "firstStartedPulling", state.firstStartedPulling, "lastFinishedPulling", state.lastFinishedPulling)
+
+		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
```

Comment:
Why record the metric from ObservedPodOnWatch, rather than in RecordStatusUpdated?  Won't that add an extra round-trip time to the status?